src/msg/async/frames_v2.cc

```diff
@@ -0,0 +1,191 @@
+#include "frames_v2.h"
+
+#include <ostream>
+
+#include <fmt/format.h>
+
+namespace ceph::msgr::v2 {
+
+// Zero-pads bufferlist to padded_len.
+static void pad_zero(bufferlist* bl, uint32_t padded_len) {
+  ceph_assert(padded_len >= bl->length());
+  if (padded_len > bl->length()) {
+    bl->append_zero(padded_len - bl->length());
+  }
+}
+
+// Discards trailing empty segments, unless there is just one segment.
+// A frame always has at least one (possibly empty) segment.
+static size_t calc_num_segments(const bufferlist* segment_bls,
+                                size_t segment_count) {
+  ceph_assert(segment_count > 0 && segment_count <= MAX_NUM_SEGMENTS);
+  for (size_t i = segment_count; i-- > 0; ) {
+    if (segment_bls[i].length() > 0) {
+      return i + 1;
+    }
+  }
+  return 1;
+}
+
+void FrameAssembler::fill_preamble(Tag tag,
+                                   preamble_block_t* preamble) const {
+  // FIPS zeroization audit 20191115: this memset is not security related.
+  ::memset(preamble, 0, sizeof(*preamble));
+
+  preamble->tag = static_cast<__u8>(tag);
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    preamble->segments[i].length = m_descs[i].logical_len;
+    preamble->segments[i].alignment = m_descs[i].align;
+  }
+  preamble->num_segments = m_descs.size();
+  preamble->crc = ceph_crc32c(
+      0, reinterpret_cast<const unsigned char*>(preamble),
+      sizeof(*preamble) - sizeof(preamble->crc));
+}
+
+uint32_t FrameAssembler::get_segment_padded_len(size_t seg_idx) const {
+  return p2roundup<uint32_t>(m_descs[seg_idx].logical_len, CRYPTO_BLOCK_SIZE);
+}
+
+uint32_t FrameAssembler::get_auth_tag_len() const {
+  return m_crypto->rx->get_extra_size_at_final();
+}
+
+// claim_append() doesn't have an rvalue reference overload
+void FrameAssembler::append_encrypt_final(bufferlist* bl) const {
+  auto tmp = m_crypto->tx->authenticated_encrypt_final();
+  bl->claim_append(tmp);
+}
+
+uint64_t FrameAssembler::get_frame_logical_len() const {
+  ceph_assert(!m_descs.empty());
+  uint64_t logical_len = 0;
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    logical_len += m_descs[i].logical_len;
+  }
+  return logical_len;
+}
+
+uint64_t FrameAssembler::get_frame_onwire_len() const {
+  ceph_assert(!m_descs.empty());
+  uint64_t onwire_len = get_preamble_onwire_len();
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    onwire_len += get_segment_onwire_len(i);
+  }
+  onwire_len += get_epilogue_onwire_len();
+  return onwire_len;
+}
+
+void FrameAssembler::asm_crc_r0(const preamble_block_t& preamble,
+                                bufferlist* segment_bls,
+                                bufferlist* frame_bl) const {
+  frame_bl->append(reinterpret_cast<const char*>(&preamble),
+                   sizeof(preamble));
+
+  epilogue_plain_block_t epilogue;
+  // FIPS zeroization audit 20191115: this memset is not security related.
+  ::memset(&epilogue, 0, sizeof(epilogue));
+
+  for (size_t i = 0; i < m_descs.size(); i++) {
```

Comment:
This assert is used to tell whether `assemble_frame()` or `disassemble_preamble()` has run, to avoid confusing users of public methods such as `get_frame_onwire_len()`.  `asm_crc_r0()` is a private helper invoked after `m_descs` is resized to a known size at the top of `assemble_frame()`, so asserting here would be bogus.

Reply:
OK, the fact they are `private` sounds convincing.